### PR TITLE
feat(amf): Add support for IPv6 pdu session procedures

### LINKED ISSUE #9103: [AMF] : Add support for IPv6 pdu session procedures
Following are the high level work items :
1. Parsing the pdu session type and requesting mobilityd for specific address (v4 or v6).
2. Store the information based on the type requested.
3. While sending to SMF the UE should be in IPAddress format.
4. Same while receiving any message from smf.
5. Paging related testing on IPv6.
6. Basic QoS sanity on IPv6 related entries.

### PR DESCRIPTION
**Pull request by Rajeshs23**
Signed-off-by: Rajeshs23 <rajesh.sure@wavelabs.ai>
## Summary
feat:
- Support of UE IPv6 pdu session procedures in amf
- Support of UE IPv4v6 pdu session procedures in amf

## Test Plan

- Verified UT cases
- Tested using ueransim
- Tested using gnbsim

Since there is no UE IPv6 support for ueransim we have kept breakpoints in gdb and externally changed pdu session type to ipv6 and ipv4v6 .Mobilityd allocated UE IPv6 and UE IPv4v6 address and given back to amf ,amf requested smf to create a pdu session with UE IPv6 and UE IPv4v6 also pdu session establishment is successful.
Tested using gnbsim and observed pdu session accept message is created with UEIPV6.
ipv6

ipv4v6typepcap
ipv6typepcap
ut_test

## Additional Information

corresponding zenhub task id #9103

### CODE REVIEW

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp [hunk 1]
```diff
@@ -71,14 +72,106 @@ static void handle_allocate_ipv4_address_status(
...
+  amf_ip_allocation_response_p->pdu_session_type = pdu_session_type;
+  amf_ip_allocation_response_p->paa.ipv6_address = in_ip6_addr;
+  amf_ip_allocation_response_p->paa.pdn_type     = IPv6;
+  amf_ip_allocation_response_p->paa.vlan         = vlan;
+
+  memcpy(
+      amf_ip_allocation_response_p->gnb_gtp_teid_ip_addr, gnb_gtp_teid_ip_addr,
+      4);
+
+  memcpy(amf_ip_allocation_response_p->apn, apn, strlen(apn) + 1);
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp:110]: memcpy third argument should be strlen(apn) -1

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp [hunk 2]
```diff
@@ -71,14 +72,106 @@ static void handle_allocate_ipv4_address_status(
...
+  amf_ip_allocation_response_p->pdu_session_type = pdu_session_type;
+  amf_ip_allocation_response_p->paa.ipv4_address = in_ip4_addr;
+  amf_ip_allocation_response_p->paa.ipv6_address = in_ip6_addr;
+  amf_ip_allocation_response_p->paa.pdn_type     = IPv4_AND_v6;
+  amf_ip_allocation_response_p->paa.vlan         = vlan;
+  memcpy(
+      amf_ip_allocation_response_p->gnb_gtp_teid_ip_addr, gnb_gtp_teid_ip_addr,
+      4);
+
+  memcpy(amf_ip_allocation_response_p->apn, apn, strlen(apn) + 1);
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp:152]: same as previous comment

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 3]
```diff
@@ -711,6 +711,52 @@ void convert_ambr(
   *ambr_value = static_cast<uint16_t>(temp_pdu_ambr_response_value);
 }
 
+// Utility function to convert address to buffer
+static int paa_to_address_info(
+    const paa_t* paa, uint8_t* pdu_address_info, uint8_t* pdu_address_length) {
+  uint32_t ip_int = 0;
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:716]: validate input pointers, add null check for arguments

Comment (ganeshg87) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:716]: add UT to validate paa_to_address_info

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp [hunk 4]
```diff
@@ -827,20 +836,61 @@ int amf_smf_handle_ip_address_response(
...
+        OAILOG_ERROR(LOG_AMF_APP, "Create IPV4 Session \n");
+      }
+    }
 
-    rc = amf_smf_create_ipv4_session_grpc_req(
-        response_p->imsi, response_p->apn, response_p->pdu_session_id,
-        response_p->pdu_session_type, response_p->gnb_gtp_teid, response_p->pti,
-        response_p->gnb_gtp_teid_ip_addr, ip_str, smf_ctx->smf_ctx_ambr);
+    if (response_p->paa.pdn_type == IPv6) {
+      char ip_v6_str[INET6_ADDRSTRLEN];
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: initialize ip_v6_str variable

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp [hunk 5]
```diff
@@ -827,20 +836,61 @@ int amf_smf_handle_ip_address_response(
...
+          response_p->pdu_session_type, response_p->gnb_gtp_teid,
+          response_p->pti, response_p->gnb_gtp_teid_ip_addr, NULL, ip_v6_str,
+          smf_ctx->smf_ctx_ambr);
+      if (rc < 0) {
+        OAILOG_ERROR(LOG_AMF_APP, "Create IPV6 Session \n");
+      }
+    }
+
+    if (response_p->paa.pdn_type == IPv4_AND_v6) {
+      char ip_v4_str[INET_ADDRSTRLEN];
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp]: initialize local variables

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPDUAddress.h [hunk 6]
```diff
@@ -14,15 +14,19 @@
 #include <cstdint>
 
 namespace magma5g {
-#define TYPE_VAL_IPV4 1
-#define IPV4_ADDRESS_LENGTH 4
+#define PDU_SESSION_TYPE_LENGTH 1
+#define PDN_SESSION_TYPE_IPV4 0b001
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPDUAddress.h]: can we use PDUSessiontype defined in M5GNasEnums.h
enum class M5GPduSessionType : uint8_t {
  IPV4         = 0b001,
  IPV6         = 0b010,
  IPV4V6       = 0b011,
  UNSTRUCTURED = 0b100,
  ETHERNET     = 0b101,
};

File: lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp [hunk 7]
```diff
@@ -754,4 +766,174 @@ TEST_F(AMFAppProcedureTest, TestPDUSession_Invalid_PDUSession_Identity) {
...
+  EXPECT_EQ(ue_context_p->amf_context.smf_ctxt_map.size(), 0);
+
+  rc = send_pdu_notification_response();
+  EXPECT_TRUE(rc == RETURNok);
+
+  /* Send uplink nas message for deregistration complete response from UE */
+  rc = send_uplink_nas_ue_deregistration_request(
+      amf_app_desc_p, ue_id, plmn, ue_initiated_dereg_hexbuf,
+      sizeof(ue_initiated_dereg_hexbuf));
+}
```
Comment (ganeshg87) [inline lte/gateway/c/core/oai/test/amf/test_amf_procedures.cpp:1147]: add validations checks number of msgs sending by AMF

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp [hunk 8]
```diff
@@ -71,14 +72,106 @@ static void handle_allocate_ipv4_address_status(
...
+  amf_ip_allocation_response_p->pdu_session_id   = pdu_session_id;
+  amf_ip_allocation_response_p->pti              = pti;
+  amf_ip_allocation_response_p->pdu_session_type = pdu_session_type;
+  amf_ip_allocation_response_p->paa.ipv6_address = in_ip6_addr;
+  amf_ip_allocation_response_p->paa.pdn_type     = IPv6;
+  amf_ip_allocation_response_p->paa.vlan         = vlan;
+
+  memcpy(
+      amf_ip_allocation_response_p->gnb_gtp_teid_ip_addr, gnb_gtp_teid_ip_addr,
+      4);
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp]: You could use gnb_gtp_teid_ip_addr_len

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp [hunk 9]
```diff
@@ -71,14 +72,106 @@ static void handle_allocate_ipv4_address_status(
...
+  amf_ip_allocation_response_p->pdu_session_id   = pdu_session_id;
+  amf_ip_allocation_response_p->pti              = pti;
+  amf_ip_allocation_response_p->pdu_session_type = pdu_session_type;
+  amf_ip_allocation_response_p->paa.ipv4_address = in_ip4_addr;
+  amf_ip_allocation_response_p->paa.ipv6_address = in_ip6_addr;
+  amf_ip_allocation_response_p->paa.pdn_type     = IPv4_AND_v6;
+  amf_ip_allocation_response_p->paa.vlan         = vlan;
+  memcpy(
+      amf_ip_allocation_response_p->gnb_gtp_teid_ip_addr, gnb_gtp_teid_ip_addr,
+      4);
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp]: You could use You could use gnb_gtp_teid_ip_addr_len.

Comment (Rajeshs23) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp]: Addressed review comment

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp (hunk 8 above)
Comment (Rajeshs23) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp]: Addressed Review comment

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp [hunk 10]
```diff
@@ -66,19 +67,111 @@ static void handle_allocate_ipv4_address_status(
...
+    uint8_t gnb_gtp_teid_ip_addr_len) {
+  MessageDef* message_p;
+  message_p =
+      itti_alloc_new_message(TASK_GRPC_SERVICE, AMF_IP_ALLOCATION_RESPONSE);
+
+  itti_amf_ip_allocation_response_t* amf_ip_allocation_response_p;
+  amf_ip_allocation_response_p = &message_p->ittiMsg.amf_ip_allocation_response;
+  memset(
+      amf_ip_allocation_response_p, 0,
+      sizeof(itti_amf_ip_allocation_response_t));
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp]: itti_alloc_new_message already returns a zeroed MessageDef in itti_alloc_new_message_sized

**Review by lionelgo — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp (hunk 10 above)
Comment (Rajeshs23) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp]: Comment addressed

Comment (Rajeshs23) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp]: Addressed review comment

**Comment by Rajeshs23** [comment]
Hi @lionelgo , I have addressed review comments. Please provide your feedback
Thank You
Rajesh

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 11]
```diff
@@ -709,18 +710,67 @@ void convert_ambr(
...
+      INT32_TO_BUFFER(ip_int, pdu_address_info);
+      *pdu_address_length = sizeof(ip_int);
+      break;
+
+    case IPv6:
+      if (paa->ipv6_prefix_length == IPV6_PREFIX_LEN) {
+        memcpy(
+            pdu_address_info, &paa->ipv6_address.s6_addr[IPV6_INTERFACE_ID_LEN],
+            paa->ipv6_prefix_length / 8);
+        *pdu_address_length = IPV6_INTERFACE_ID_LEN;
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:733]: wouldn't it be 
memcpy(
            pdu_address_info, &paa->ipv6_address.s6_addr[IPV6_PREFIX_LEN/8],
            IPV6_INTERFACE_ID_LEN);
?

Comment (lionelgo) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:733]: Just a comment that we can skip here but what seems confusing is that PREFIX unit is in bits and interface ID in bytes , would be good to harmonize or add a suffix ?

**Review by lionelgo — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp (hunk 11 above)
Comment (Rajeshs23) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp:733]: Addressed the review comment

**Review by lionelgo — APPROVED** [review]

File: lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp (hunk 1 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp:110]: how is this resolved? why is it `strlen(apn) + 1`?

Comment (ulaskozat) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp:110]: cc @ganeshg87

Comment (Rajeshs23) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp:110]: memcpy(dest, src, strlen(src)-1);
-> Output is coming as : "interne"



memcpy(dest, src, strlen(src)+1);
-> Will give the correct output : "internet"
Also including the Null Termination



Comment clarification was provided and agreed.

Comment (ganeshg87) [reply lte/gateway/c/core/oai/lib/n11/M5GMobilityServiceClient.cpp:110]: the third argument of memcpy is "numBytes" bytes need copy from  "src" address (string length + 1) correct.

**Review by ganeshg87 — APPROVED** [review]
lgtm..

**Pull request merged**